Docs & tests for rejectWithValue (#2)
phryneas/redux-toolkit: docs/api/createAsyncThunk.md

### `payloadCreator`

> A callback function that should return a promise containing the result of some asynchronous logic. It may also return a value synchronously. If there is an error, it should return a rejected promise containing one of the following: an `Error` instance, a plain value such as a descriptive error message, or a manually rejected error with a defined payload.

Reviewer: ```suggestion
A callback function that should return a promise containing the result of some asynchronous logic. It may also return a value synchronously. If there is an error, it should either return a rejected promise containing with an `Error` instance or a plain value such as a descriptive error message or otherwise a resolved promise with a `RejectWithValue` argument as returned by the `thunkApi.rejectWithValue` function.
```

Might still be worded better, but in the third case, it's not a rejected promise, but a resolved one.

- call the `payloadCreator` callback and wait for the returned promise to settle
- when the promise settles:
  - if the promise resolved successfully, dispatch the `fulfilled` action with the promise value as `action.payload`

>   - if the promise resolved successfully or failed, but was returned with `rejectWithValue(value)`, dispatch the `rejected` action with the value passed into `action.payload` and 'Rejected' as `action.error.message`

Reviewer: ```suggestion
  - if the promise resolved with a `rejectWithValue(value)` return value, dispatch the `rejected` action with the value passed into `action.payload` and 'Rejected' as `action.error.message`
```

  }
}

> interface RejectedWithValueAction<ThunkArg, RejectedValue> {

Reviewer: This is not really an existing type with that name, I'm not sure if we should document it that way.

Reviewer: To be fair: none of these types actually exist. Should we have nonexistent interfaces interfaces in the documentation?

Reviewer 2: There's certainly been times when I've written kinda hand-wave-y types in docs to illustrate a point, but if we do that it probably needs to be clarified.

Author: I'm not sure of the best language to use here. Open to suggestions.

    }
  }
)

const usersSlice = createSlice({
  name: 'users',
  initialState: {
    entities: {},
    loading: 'idle',
    error: null
  },
  reducers: {},
  extraReducers: {
    [updateUser.fullfilled]: (state, action) => {
      const user = action.payload
      state.entities[user.id] = user
    },
    [updateUser.rejected]: (state, action) => {
      if (action.payload) {
        // if a rejected action has a payload, it means that it was returned with rejectWithValue
        state.error = action.payload.errorMessage
      } else {
        state.error = action.error
      }
    }
  }
})

const UsersComponent = () => {
  const { users, loading, error } = useSelector(state => state.users)
  const dispatch = useDispatch()

  const updateUser = async userData => {
    const resultAction = await dispatch(updateUser(userData))
    if (updateUser.fulfilled.match(resultAction)) {
      const user = unwrapResult(resultAction)
      showToast('success', `Updated ${user.name}`)
    } else {
      if (resultAction.payload) {
        // This is assuming the api returned a 400 error with a body of { errorMessage: 'Validation errors', field_errors: [{ field_name: 'Should be a string' }]}

>         setErrors(resultAction.payload.field_errors)

Reviewer 2: I realize this is a fictional example, but what's `setErrors` supposed to be?

Author: The intent was to help identify a common use from a form library (in this case, Formik). Most form libs have some type of utility like `setErrors`. I tried to make that clear with the comment above saying it was validation errors that happened when updating user details. Looking at this, the example field_errors is wrong anyways and I'll correct if you don't mind this staying in. (should be `field_errors: {field:'error', other_field: 'error'}`)

Author: Note added about where `setErrors` comes from

      } else {
        showToast('error', `Update failed: ${resultAction.error}`)
      }
    }
  }

  // render UI here
}
```

> - TypeScript: Using rejectWithValue to access a custom rejected payload in a component

Reviewer: This seems almost tutorial-level complex. Maybe put a "dumbed down" (I dumbed down that one a little for a type-test) variant of this onto the "usage with TS" page and keep this one for a tutorial page?

Author: I did add some basic usage in the usage-with-typescript guide that covers everything this example shows. I'll add an additional comment that addresses what the `setErrors` portion does here.

>   _Note: this is a contrived example assuming our userAPI only ever throws a validation-specific errors_

Reviewer 2: ```suggestion
  _Note: this is a contrived example assuming our userAPI only ever throws validation-specific errors_
```

```typescript
import { createAsyncThunk, createSlice, unwrapResult } from '@reduxjs/toolkit'
import { userAPI } from './userAPI'

// Sample types that will be used
interface User {
  first_name: string
  last_name: string
  email: string
}

interface ValidationErrors {
  errorMessage: string
  field_errors: Record<string, string>
}

interface UpdateUserResponse {
  user: User
  success: boolean
}

const updateUser = createAsyncThunk<

>   User,

Author: In general, when looking at the types on this - `Returned` always seems to be shown as `Promise<Returned>`. Even if you do something like:

```
createAsyncThunk('test/thing', () => 'hello')
```
This will be a promise, regardless. My point is that we'd never need to type `createAsyncThunk<Promise<Collection>>()`, so we should probably remove it from the example docs of `createAsyncThunk<Promise<MyData[]>()`. Or is this incorrect?

Reviewer: Yeah, that's true. :+1:

  Partial<User>,
  {
    rejectValue: ValidationErrors
  }
>('users/update', async (userData, { rejectWithValue }) => {
  try {
    const { id, ...data } = userData
    const response = await userAPI.updateById<UpdateUserResponse>(id, data)
    return response.data.user
  } catch (err) {
    let error: AxiosError<ValidationErrors> = err // cast the error for access
    if (!error.response) {
      throw err
    }
    // We got validation errors, let's return those so we can reference in our component and set form errors
    return rejectWithValue(error.response.data)
  }
})

interface UsersState {
  error: string | null
  entities: Record<string, User>
}

const initialState: UsersState = {
  entities: {},
  error: null
}

const usersSlice = createSlice({
  name: 'users',
  initialState,
  reducers: {},

>   extraReducers: builder => {

Reviewer 2: Might want to add a descriptive comment in here saying we're using the `builder` callback form to get correct typing of the reducers from the action creators.

    builder.addCase(updateUser.fulfilled, (state, { payload }) => {
      state.entities[payload.id] = payload
    })
    builder.addCase(updateUser.rejected, (state, action) => {
      if (action.payload) {
        // Being that we passed in ValidationErrors to rejectType in `createAsyncThunk`, the payload will be available here.
        state.error = action.payload.errorMessage
      } else {
        state.error = action.error
      }
    })
  }
})

const UsersComponent = () => {

>   const { users, loading, error } = useSelector(state => state.users)

Reviewer 2: If this is a TS example, we'd need to type `state` here: 
```suggestion
  const { users, loading, error } = useSelector( (state: RootState) => state.users)
```

Author: good catch!
